keanu-project/src/main/java/io/improbable/keanu/util/dot/WriteDot.java

```diff
@@ -0,0 +1,279 @@
+            writer.write("digraph BayesianNetwork {\n");
+
+            // Iterate over vertices and obtain the necessary label and conneciton info.
+            obtainGraphInfo(vertex, degree);
+
+            // Write out labels and connections.
+            outputConnections(verticeIdsToConnectionInfo.values());
+            outputLabels();
+
+            writer.write("}");
+            writer.close();
+        }
+        catch (IOException e) {
+            System.out.println("Encountered an issue writing to the specified file (" + fileName + "):");
+            e.printStackTrace();
+        }
+    }
+
+
+    // Make sure the output directory exists and that no files are being overwritten.
+    private static File getOutputFile(String fileName) {
+
+        File outputFile = new File(fileName);
+
+        // Make sure the output directory exists.
+        if (outputFile.getParentFile() != null && !outputFile.getParentFile().exists()){
+            outputFile.getParentFile().mkdirs();
+        }
+
+        // Add an index to the filename if a file with the specified name already exists.
+        if (outputFile.exists()) {
+            String baseName = FilenameUtils.getBaseName(fileName);
+            String extension = FilenameUtils.getExtension(fileName);
+            int counter = 1;
+            while(outputFile.exists())
+            {
+                outputFile = new File(outputFile.getParent(), baseName + "_" + (counter++) + "." + extension);
+            }
+        }
+        return outputFile;
+    }
+
+
+    /**
+     * Iterates over the vertices of the graph, storing information about vertex labels to display and connections between vertices.
+     *
+     * @param vertex vertex around which the graph will be visualised
+     * @param degree degree of connections to be visualised; for instance, if the degree is 1,
+     *               only connections between the vertex amd it's parents and children will be written out to the DOT file.
+     */
+    private static void obtainGraphInfo(Vertex vertex, int degree) {
+
+        HashSet<VertexId> processedVertices = new HashSet<>();
+        HashSet<Vertex> verticesToProcessNow = new HashSet<>();
+        verticesToProcessNow.add(vertex);
+
+        // Iterate over the graph till the specified degree or till there are no more unprocessed vertices left.
+        int iterationIndex = 0;
+        while(iterationIndex < degree && !verticesToProcessNow.isEmpty()) {
+
+            HashSet<Vertex> verticesToProcessNext = new HashSet<>();
+
+            for (Vertex v : verticesToProcessNow) {
+
+                // Use hashed IDs as the unique identifier for each vertex and store the other information in labels.
+                VertexId vId = v.getId();
+                setLabel(v);
+
+                // Iterate over all vertices connected to this one.
+                Stream<Vertex> connectedVertices = Stream.concat(v.getParents().stream(), v.getChildren().stream());
```

Comment:
There are quite a few implementations of graph walkers in the codebase already - perhaps we should make a single implementation that all the code can share?

Reply:
Can we chat about this in person? You guys definitely know the code base better than I do, but I couldn't find anything quite like what I needed.